include/envoy/upstream/upstream.h

```diff
@@ -380,6 +380,19 @@ class PrioritySet {
                            LocalityWeightsConstSharedPtr locality_weights,
                            const HostVector& hosts_added, const HostVector& hosts_removed,
                            absl::optional<uint32_t> overprovisioning_factor) PURE;
+
+  typedef std::function<void(uint32_t, UpdateHostsParams&&, LocalityWeightsConstSharedPtr,
```

Comment:
I like this idea as well. My main feedback is that I think this would benefit from adding in C++ interfaces, as the callback/continuation style is a bit complicated to follow, and we can add decent documentation (Doxygen-style) on the interface classes.